sui-execution/latest/sui-adapter/src/package_unification.rs

```diff
@@ -0,0 +1,379 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use move_binary_format::{binary_config::BinaryConfig, file_format::Visibility};
+use sui_types::{
+    base_types::{ObjectID, SequenceNumber},
+    move_package::{normalize_module, MovePackage},
+    storage::BackingPackageStore,
+    transaction::{Command, ProgrammableTransaction},
+    type_input::TypeInput,
+    Identifier,
+};
+
+pub struct PTBLinkageMetadata {
+    pub types: Vec<ObjectID>,
+    pub entry_functions: Vec<ObjectID>,
+    pub non_entry_functions: Vec<ObjectID>,
+    pub publication_linkages: Vec<ObjectID>,
+    pub upgrading_packages: Vec<ObjectID>,
+    pub all_packages: BTreeMap<ObjectID, MovePackage>,
+}
+
+pub struct LinkageConfig {
+    pub fix_top_level_functions: bool,
+    pub fix_types: bool,
+}
+
+#[derive(Debug)]
+pub enum ConflictResolution {
+    Exact(SequenceNumber, ObjectID),
+    AtLeast(SequenceNumber, ObjectID),
+    Never(ObjectID),
+}
+
+impl LinkageConfig {
+    pub fn strict() -> Self {
+        Self {
+            fix_top_level_functions: true,
+            fix_types: true,
+        }
+    }
+
+    pub fn loose() -> Self {
+        Self {
+            fix_top_level_functions: false,
+            fix_types: false,
+        }
+    }
+}
+
+impl ConflictResolution {
+    pub fn unify(&mut self, other: &ConflictResolution) -> anyhow::Result<()> {
+        match (&self, other) {
+            // If we ever try to unify with a Never we fail.
+            (ConflictResolution::Never(_), _) | (_, ConflictResolution::Never(_)) => {
+                return Err(anyhow::anyhow!(
+                    "Cannot unify with Never: {:?} and {:?}",
+                    self,
+                    other
+                ));
+            }
+            // If we have two exact resolutions, they must be the same.
+            (ConflictResolution::Exact(sv, self_id), ConflictResolution::Exact(ov, other_id)) => {
+                if self_id != other_id {
```

Comment:
I'm confused. Why aren't we checking the `sv` and `ov`?

Reply:
Is it because if the ids are the same, the sequence number should also be the same?

Reply:
Correct. But it doesn't hurt to double-check.